### PR TITLE
Add Medication and MedicationList

### PR DESCRIPTION
**Pull request by BlopApple**

### CODE REVIEW

File: src/main/java/seedu/uninurse/model/medication/exceptions/DuplicateMedicationException.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package seedu.uninurse.model.medication.exceptions;
+
+/**
+ * Signals that the operation will result in duplicate Conditions.
+ * Conditions are considered duplicates if they have the same name.
```
Comment (izzahaj) [inline src/main/java/seedu/uninurse/model/medication/exceptions/DuplicateMedicationException.java:5]: I think you forgot to change the JavaDoc comments
```suggestion
 * Medications are considered duplicates if they have the same name.
```

File: src/main/java/seedu/uninurse/model/medication/exceptions/DuplicateMedicationException.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package seedu.uninurse.model.medication.exceptions;
+
+/**
+ * Signals that the operation will result in duplicate Conditions.
```
Comment (izzahaj) [inline src/main/java/seedu/uninurse/model/medication/exceptions/DuplicateMedicationException.java:4]:
```suggestion
 * Signals that the operation will result in duplicate Medications.
```

File: src/main/java/seedu/uninurse/model/medication/exceptions/UnmodifiedMedicationException.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package seedu.uninurse.model.medication.exceptions;
+
+/**
+ * Signals that the operation will result in duplicate Conditions.
```
Comment (izzahaj) [inline src/main/java/seedu/uninurse/model/medication/exceptions/UnmodifiedMedicationException.java:4]:
```suggestion
 * Signals that the operation will result in duplicate medications.
```

File: src/main/java/seedu/uninurse/model/medication/exceptions/UnmodifiedMedicationException.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+ * Conditions are considered duplicates if they have the same name.
```
Comment (izzahaj) [inline src/main/java/seedu/uninurse/model/medication/exceptions/UnmodifiedMedicationException.java:5]:
```suggestion
 * Medications are considered duplicates if they have the same name.
```

File: src/main/java/seedu/uninurse/model/medication/MedicationList.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+    @Override
+    public boolean isEmpty() {
+        return internalMedicationList.isEmpty();
+    }
+
+    @Override
+    public List<Medication> getInternalList() {
+        // returns a copy of the internal list to prevent modification to original one
+        return new ArrayList<>(internalMedicationList);
```
Comment (izzahaj) [inline src/main/java/seedu/uninurse/model/medication/MedicationList.java:96]: I made this change in my newer version of `ConditionList` so there's a stronger notion of immutability. So far no other external methods require modifying the internal list (even if there the list is copy).
```suggestion
        return Collections.unmodifiableList(internalMedicationList);
```

**Comment by BlopApple** [comment]
Yup I had a look at your suggestions and rectified them accordingly.

File: src/main/java/seedu/uninurse/model/medication/MedicationList.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+import java.util.List;
+
+import seedu.uninurse.model.GenericList;
+import seedu.uninurse.model.medication.exceptions.DuplicateMedicationException;
+import seedu.uninurse.model.medication.exceptions.MedicationNotFoundException;
+import seedu.uninurse.model.medication.exceptions.UnmodifiedMedicationException;
+
+/**
+ * Represents a unique list of medications for a particular patient.
+ * Two medications with the same name are considered duplicates and are hence, not allowed.
```
Comment (izzahaj) [inline src/main/java/seedu/uninurse/model/medication/MedicationList.java:16]:
```suggestion
 * Two medications with the same name and dosage type are considered duplicates and are hence, not allowed.
```

**Review by izzahaj — CHANGES_REQUESTED** [review]
Looks ok, just one small JavaDoc inconsistency regarding what are considered duplicate medications.

File: src/main/java/seedu/uninurse/model/medication/MedicationList.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        final StringBuilder builder = new StringBuilder();
+        internalMedicationList.forEach(m -> {
+            int index = internalMedicationList.indexOf(m);
+            if (index == 0) {
+                builder.append(index + 1)
+                        .append(". ")
+                        .append(m);
+            } else {
+                builder.append("\n")
+                        .append(index + 1)
```
Comment (izzahaj) [inline src/main/java/seedu/uninurse/model/medication/MedicationList.java:110]: oops also a bug I found in my own branch
```suggestion
                        .append(index + 1)
                        .append(". ")
```

File: src/main/java/seedu/uninurse/model/medication/MedicationList.java (hunk 6 above)
Comment (BlopApple) [reply src/main/java/seedu/uninurse/model/medication/MedicationList.java]: Yup I updated it.

File: src/main/java/seedu/uninurse/model/medication/MedicationList.java (hunk 7 above)
Comment (BlopApple) [reply src/main/java/seedu/uninurse/model/medication/MedicationList.java:110]: Yup I amended it.

**Review by izzahaj — APPROVED** [review]
LGTM

**Pull request merged**